OpenRA.Mods.Common/Activities/Air/ResupplyAircraft.cs

```diff
@@ -16,59 +16,40 @@
 
 namespace OpenRA.Mods.Common.Activities
 {
-	public class ResupplyAircraft : Activity
+	public class ResupplyAircraft : CompositeActivity
 	{
-		readonly Aircraft aircraft;
-		Activity inner;
+		public ResupplyAircraft(Actor self) { }
 
-		public ResupplyAircraft(Actor self)
+		protected override void OnFirstRun(Actor self)
 		{
-			aircraft = self.Trait<Aircraft>();
-		}
+			var aircraft = self.Trait<Aircraft>();
+			var host = aircraft.GetActorBelow();
 
-		public override Activity Tick(Actor self)
-		{
-			if (IsCanceled)
-				return NextActivity;
+			if (host == null)
+				return;
 
-			if (inner == null)
+			if (aircraft.IsPlane)
 			{
-				var host = aircraft.GetActorBelow();
-
-				if (host == null)
-					return NextActivity;
-
-				if (aircraft.IsPlane)
-				{
-					inner = ActivityUtils.SequenceActivities(
-						aircraft.GetResupplyActivities(host)
-						.Append(new AllowYieldingReservation(self))
-						.Append(new WaitFor(() => NextActivity != null || aircraft.ReservedActor == null))
-						.ToArray());
-				}
-				else
-				{
-					// Helicopters should take off from their helipad immediately after resupplying.
-					// HACK: Append NextActivity to TakeOff to avoid moving to the Rallypoint (if NextActivity is non-null).
-					inner = ActivityUtils.SequenceActivities(
-						aircraft.GetResupplyActivities(host)
-						.Append(new AllowYieldingReservation(self))
-						.Append(new TakeOff(self)).Append(NextActivity).ToArray());
-				}
+				ChildActivity = ActivityUtils.SequenceActivities(
+					aircraft.GetResupplyActivities(host)
+					.Append(new AllowYieldingReservation(self))
+					.Append(new WaitFor(() => nextActivity != null || aircraft.ReservedActor == null))
+					.ToArray());
 			}
 			else
-				inner = ActivityUtils.RunActivity(self, inner);
-
-			// The inner == NextActivity check is needed here because of the TakeOff issue mentioned in the comment above.
-			return inner == null || inner == NextActivity ? NextActivity : this;
+			{
+				// Helicopters should take off from their helipad immediately after resupplying.
+				// HACK: Append nextActivity to TakeOff to avoid moving to the Rallypoint (if nextActivity is non-null).
+				ChildActivity = ActivityUtils.SequenceActivities(
+					aircraft.GetResupplyActivities(host)
+					.Append(new AllowYieldingReservation(self))
+					.Append(new TakeOff(self)).Append(nextActivity).ToArray());
+			}
 		}
 
-		public override bool Cancel(Actor self)
+		public override Activity Tick(Actor self)
 		{
-			if (!IsCanceled && inner != null && !inner.Cancel(self))
-				return false;
-
-			return base.Cancel(self);
+			return NextActivity;
```

Comment:
We definitely need an obvious comment, or maybe a rethink of `CompositeActivity` here.

My first assumption from reading this class is that `OnFirstRun` is queuing up the child activities that are then ticked by the actor code in preference to this one.  Only once they have all completed does this parent activity Tick, and, with nothing left to do, moves on to the next logical activity.  This assumption is completely wrong.